mutiny-core/src/redshift.rs

```diff
@@ -273,57 +273,65 @@ impl RedshiftManager for NodeManager {
                 nodes
                     .get(&new_receiving_node)
                     .ok_or(MutinyError::NotFound)?
             }
         };
 
-        // TODO for loop while max_sats is not hit
-        // TODO get the real number to attempt
-        let local_max = max_sats;
-
-        // get an invoice from the receiving node
-        let invoice = receiving_node
-            .create_invoice(Some(local_max), None, None)
-            .await?; // TODO probably should handle error
-
-        // make attempts to pay it
-        match sending_node
-            .pay_invoice_with_timeout(&invoice, None, None)
-            .await
-        {
-            Ok(i) => {
-                if i.paid {
-                    debug!("paid the redshift invoice");
-                    rs.payment_attempted(local_max);
-
-                    // TODO when running in a for loop, only
-                    // return this when it is actually done
-                    rs.status = RedshiftStatus::Completed;
-                } else {
-                    // TODO need to handle payments still pending
-                    debug!("payment still pending...");
+        // attempt payments in loop until we sent all or hit min sats
+        let mut local_max_sats = max_sats;
+        rs.status = RedshiftStatus::AttemptingPayments(0);
+        loop {
+            // keep trying until the amount is too small to send through
+            if local_max_sats < min_sats {
+                // if no payments were made, consider it a fail
+                if let RedshiftStatus::AttemptingPayments(0) = rs.status {
+                    rs.fail("exhausted all payment attempt amounts".to_string());
                 }
+                break;
             }
-            Err(e) => {
-                // TODO keep going through loop
-                error!("could not pay: {e}");
-                rs.status = RedshiftStatus::Failed("could not pay invoice".to_string());
-                self.storage.update_redshift(rs)?;
-                return Err(MutinyError::Other(anyhow!(
-                    "could not redshift pay invoice"
-                )));
+
+            // get an invoice from the receiving node
+            let invoice = receiving_node
+                .create_invoice(Some(local_max_sats), None, None)
+                .await?; // TODO probably should handle error
+
+            // make attempts to pay it
+            match sending_node
+                .pay_invoice_with_timeout(&invoice, None, None)
+                .await
+            {
+                Ok(i) => {
+                    if i.paid {
+                        debug!("paid the redshift invoice");
+                        rs.payment_attempted(local_max_sats);
+
+                        // check if the max amount was sent on first try
+                        if local_max_sats == max_sats {
+                            rs.status = RedshiftStatus::Completed;
+                            break;
+                        }
```

Comment:
also need an exit condition for if the first fails but a later attempt succeeds

Reply:
With the strategy you're doing, probably fine to just exit if any payment works

Reply:
It should now consider the previous amount paid when determining if a successful payment has now maxed out the channel. Otherwise, the amount will keep going down until it hits the min amount.